### PR TITLE
feat: localhost subdomain redirection

### PR DESCRIPTION
**Pull request by tomicvladan**
#96 #97 #98

### CODE REVIEW

**Review by nugaon — COMMENTED** [review]
I just tried it out with my local build but it does not redirect the `bzz://{content hash}` to the localhost subdomain...

**Comment by tomicvladan** [comment]
Now it should work. Could you try again?

**Comment by nugaon** [comment]
one additional thing that we missed out is updating `sandboxListener` in the `bee-api.listener.ts`.
We don't need to apply sandbox on bee responses if the subdomain redirection is used.

**Review by nugaon — CHANGES_REQUESTED** [review]
it still does not work on my side...
we need to redirect
- `bzz://{swarm hash or ens}/{path}` and
- `https://{cid or ens}.bzz.link/{path}`
to `http://{cid or ens without .eth}.swarm.localhost:{port}/{path}`

**Comment by nugaon** [comment]
now it seems like it creates a CID from a CID of the `bzz.link` address.
e.g. `bah5acgzaunpsiqopkg7tmibpy2g3jchobzsjinftr4ps7hicxgvjstflnhtq.bzz.link` -> `http://a35f2441cf51bf36202fc68db488ee0e649434b38f1f2f9d02b9aa994cab69e7.swarm.localhost:1633/`

**Comment by nugaon** [comment]
now, the subdomain is mapped to the swarm hash reference instead of the CID. please check out the desired mapping in [my previous comment](https://github.com/ethersphere/swarm-extension/pull/136#pullrequestreview-1102469999)

File: src/background/listener/bee-api.listener.ts [hunk 1]
```diff
@@ -149,7 +149,7 @@ export class BeeApiListener {
         const bzzReference = subdomainToBzzResource(subdomain) + pathWithParams
         console.log('bzz link redirect', bzzReference, url, pathWithParams)
 
-        this.redirectToBzzReference(bzzReference, tabId)
+        this.redirectToBzzReference(bzzReference, tabId, true)
```
Comment (nugaon) [inline src/background/listener/bee-api.listener.ts:152]: why do you do that? `cid.bzz.link` also should be mapped to `{cid}.swarm.localhost`

File: src/background/listener/bee-api.listener.ts [hunk 2]
```diff
@@ -276,8 +276,25 @@ export class BeeApiListener {
...
+      url = `${this._beeApiUrl}/bzz/${bzzReference}`
+    } else {
+      const [hash, path] = bzzReference.split(/\/(.*)/s)
+      let subdomain = hash
+
+      if (subdomain.endsWith('.eth')) {
+        subdomain = subdomain.substring(0, subdomain.length - 4)
+      }
+
+      url = createSubdomainUrl(this._beeApiUrl, subdomain)
```
Comment (nugaon) [inline src/background/listener/bee-api.listener.ts:292]: here the subdomain is a swarm hash, but u don't transform it into a CID that the swarm.localhost subdomain requires.

File: src/background/listener/bee-api.listener.ts [hunk 3]
```diff
@@ -276,8 +276,25 @@ export class BeeApiListener {
    * @param bzzReference in form of $ROOT_HASH<$PATH><$QUERY>
    * @param tabId the tab will be navigated to the dApp page
    */
-  private redirectToBzzReference(bzzReference: string, tabId: number) {
-    const url = `${this._beeApiUrl}/bzz/${bzzReference}`
+  private redirectToBzzReference(bzzReference: string, tabId: number, preventSubdomainRedirection = false) {
```
Comment (nugaon) [inline src/background/listener/bee-api.listener.ts:279]: remove `preventSubdomainRedirection
```suggestion
  private redirectToBzzReference(bzzReference: string, tabId: number) {
```

**Review by nugaon — APPROVED** [review]
LGTM!

**Pull request merged**